miditok/tokenizations/remi_plus.py

```diff
@@ -0,0 +1,814 @@
+    """
+
+    def __init__(
+        self,
+        pitch_range: range = PITCH_RANGE,
+        beat_res: Dict[Tuple[int, int], int] = BEAT_RES,
+        nb_velocities: int = NB_VELOCITIES,
+        additional_tokens: Dict[
+            str, Union[bool, int, Tuple[int, int]]
+        ] = ADDITIONAL_TOKENS,
+        special_tokens: List[str] = SPECIAL_TOKENS,
+        params: Optional[Union[str, Path]] = None,
+        num_bars: Optional[int] = None,
+    ):
+        self.encoder = []
+        additional_tokens["Program"] = True  # required
+        additional_tokens["Rest"] = False
+        self.num_bars = num_bars
+        super().__init__(
+            pitch_range,
+            beat_res,
+            nb_velocities,
+            additional_tokens,
+            special_tokens,
+            unique_track=True,  # handles multi-track sequences in single stream
+            params=params,  # type: ignore
+        )
+
+    def __notes_to_events(
+        self, notes_with_program: List[Tuple[Note, Tuple[int, bool]]]
+    ) -> List[Event]:
+        """Convert multi track notes into one Token sequence.
+
+        :param notes_with_program: List[Tuple[Note, program]]
+                Note is a instance of `miditoolkit.Note`, and the program
+                is a tuple of (program_number: int, is_drum: bool).
+        :return: sequences of Event.
+        """
+        # Make sure the notes are sorted first by their onset (start) times, second by pitch
+        # notes.sort(key=lambda x: (x.start, x.pitch))  # done in midi_to_tokens
+        time_division = self._current_midi_metadata["time_division"]
+        ticks_per_sample = time_division / max(self.beat_res.values())
+        ticks_per_bar = self._current_midi_metadata["time_division"] * 4
+        dur_bins = self._durations_ticks[self._current_midi_metadata["time_division"]]
+        # Creates events
+        events: List[Event] = []
+        previous_tick = -1
+        previous_note_end = (
+            notes_with_program[0][0].start + 1
+        )  # so that no rest is created before the first note
+        current_bar = -1
+        current_tempo_idx = 0
+        current_tempo = self._current_midi_metadata["tempo_changes"][
+            current_tempo_idx
+        ].tempo
+        current_time_sig_idx = 0
+        current_time_sig_tick = 0
+        current_time_sig_bar = 0
+        time_sig_change = self._current_midi_metadata["time_sig_changes"][
+            current_time_sig_idx
+        ]
+        current_time_sig = self._reduce_time_signature(
+            time_sig_change.numerator, time_sig_change.denominator
+        )
+        # Run chord extraction for whole note sequences before tokenization
+        current_chord_idx = 0
+        current_chord = ""  # e.g. C#:min/A
+        if self.additional_tokens.get("Chord", False):  # "Chord" in additional tokens
+            chord_results = REMIPlusChord.extract([np[0] for np in notes_with_program])
```

Comment:
Shadows name 'np' from outer scope
This creates warning in th same method l514 when using numpy
Rename it ?